lib/ansible/cli/galaxy.py

```diff
@@ -969,39 +991,115 @@ def _display_role(gr):
-        else:
-            # show all valid roles in the roles_path directory
-            roles_path = context.CLIARGS['roles_path']
-            path_found = False
-            warnings = []
-            for path in roles_path:
-                role_path = os.path.expanduser(path)
-                if not os.path.exists(role_path):
-                    warnings.append("- the configured path %s does not exist." % role_path)
-                    continue
-                elif not os.path.isdir(role_path):
-                    warnings.append("- the configured path %s, exists, but it is not a directory." % role_path)
-                    continue
-                display.display('# %s' % role_path)
-                path_files = os.listdir(role_path)
+                # show all valid roles in the roles_path directory
+                roles_path = context.CLIARGS['roles_path']
+                for path in roles_path:
+                    role_path = os.path.expanduser(path)
+                    if not os.path.exists(role_path):
+                        warnings.append("- the configured path %s does not exist." % role_path)
+                        continue
+                    elif not os.path.isdir(role_path):
+                        warnings.append("- the configured path %s, exists, but it is not a directory." % role_path)
+                        continue
+                    display.display('# %s' % role_path)
+                    path_files = os.listdir(role_path)
+                    path_found = True
+                    for path_file in path_files:
+                        gr = GalaxyRole(self.galaxy, self.api, path_file, path=path)
+                        if gr.metadata:
+                            _display_role(gr)
+
+        elif context.CLIARGS['type'] == 'collection':
+            collections_search_paths = context.CLIARGS['collections_path']
+            collection_name = context.CLIARGS['collection']
+
+            # TODO: Should we support listing all collections in a given namespace?
+
+            # list a specific collection
+            if collection_name:
+                for path in collections_search_paths:
+                    collection_path = GalaxyCLI._resolve_path(path)
+                    if not os.path.exists(path):
+                        if path in C.COLLECTIONS_PATHS:
+                            # don't warn for missing default dirs
+                            continue
+                        warnings.append("- the configured path {0} does not exist.".format(collection_path))
+                        continue
+                    elif not os.path.isdir(collection_path):
+                        warnings.append("- the configured path {0}, exists, but it is not a directory.".format(collection_path))
+                        continue
+
+                    path_found = True
+                    display.display("# {0}".format(path))
+
+                    # Make sure we look inside the 'ansible_collections' dir
```

Comment:
This seems like a weird dance. Is there a better way to do this? Took me a while to figure out why I was having trouble getting at the actual collection directory.

Reply:
I also had to do this in the verify PR. I'm not sure of a better way. `install` also does it. Maybe we should add a staticmethod below _resolve_path to also append ansible_collections (or possibly incorporate it in _resolve_path?) to reduce the duplicate code.

Reply:
Added a new function in #66441 to handle this so we can both use it.